pkg/networkservice/acl/acl_parser.go

```diff
@@ -0,0 +1,273 @@
+	"esp":    ip_types.IP_API_PROTO_ESP,
+	"ah":     ip_types.IP_API_PROTO_AH,
+	"icmp6":  ip_types.IP_API_PROTO_ICMP6,
+	"eigrp":  ip_types.IP_API_PROTO_EIGRP,
+	"ospf":   ip_types.IP_API_PROTO_OSPF,
+	"sctp":   ip_types.IP_API_PROTO_SCTP,
+}
+
+func addProtocol(parsed map[string]string, rule *acl_types.ACLRule) error {
+	proto, ok := parsed[protocol]
+	if !ok {
+		return errors.New("protocol is missing")
+	}
+
+	p, ok := protocolMap[proto]
+	if !ok {
+		return errors.New("invalid protocol")
+	}
+
+	rule.Proto = p
+	return nil
+}
+
+func addSrcPortOrIcmpType(parsed map[string]string, rule *acl_types.ACLRule) error {
+	if rule.Proto == ip_types.IP_API_PROTO_ICMP || rule.Proto == ip_types.IP_API_PROTO_ICMP6 {
+		first, last, err := findNumberPairByKeys(parsed, icmpTypeFirst, icmpTypeLast)
+		if err != nil {
+			return err
+		}
+
+		rule.SrcportOrIcmptypeFirst = first
+		rule.SrcportOrIcmptypeLast = last
+
+		return nil
+	}
+
+	first, last, err := findNumberPairByKeys(parsed, srcPortFirst, srcPortLast)
+	if err != nil {
+		return err
+	}
+
+	rule.SrcportOrIcmptypeFirst = first
+	rule.SrcportOrIcmptypeLast = last
+
+	return nil
+}
+
+func addDstPortOrIcmpCode(parsed map[string]string, rule *acl_types.ACLRule) error {
+	if rule.Proto == ip_types.IP_API_PROTO_ICMP || rule.Proto == ip_types.IP_API_PROTO_ICMP6 {
+		first, last, err := findNumberPairByKeys(parsed, icmpCodeFirst, icmpCodeLast)
+		if err != nil {
+			return err
+		}
+
+		rule.DstportOrIcmpcodeFirst = first
+		rule.DstportOrIcmpcodeLast = last
+
+		return nil
+	}
+
+	first, last, err := findNumberPairByKeys(parsed, dstPortFirst, dstPortLast)
+	if err != nil {
+		return err
+	}
+
+	rule.DstportOrIcmpcodeFirst = first
+	rule.DstportOrIcmpcodeLast = last
+
+	return nil
+}
+
+func addTCPFlags(parsed map[string]string, rule *acl_types.ACLRule) {
+	tcpFlagsMask, ok := parsed[tcpFlagsMask]
+	if !ok {
+		return
+	}
+	tcpFlagsVal, ok := parsed[tcpFlagsValue]
+	if !ok {
+		return
+	}
+
+	mask, err := strconv.Atoi(tcpFlagsMask)
+	if err != nil {
+		return
+	}
+	val, err := strconv.Atoi(tcpFlagsVal)
+	if err != nil {
+		return
+	}
+
+	rule.TCPFlagsMask = uint8(mask)
+	rule.TCPFlagsValue = uint8(val)
+}
+
+// parseKVStringToMap parses the input string "ke1${kvsep}val1${sep}key2${kvsep}val2${sep}..." to map
+func parseKVStringToMap(input, sep, kvsep string) map[string]string {
+	result := map[string]string{}
+	pairs := strings.Split(input, sep)
+	for _, pair := range pairs {
+		k, v := parseKV(pair, kvsep)
+		result[k] = v
+	}
+	return result
+}
+
+func parseKV(kv, kvsep string) (key, val string) {
+	keyValue := strings.Split(kv, kvsep)
+	if len(keyValue) != 2 {
+		keyValue = []string{"", ""}
+	}
+	return strings.Trim(keyValue[0], " "), strings.Trim(keyValue[1], " ")
+}
+
+func findNumberPairByKeys(parsed map[string]string, key1, key2 string) (fisrtval, lastval uint16, err error) {
+	first, ok := parsed[key1]
+	if !ok {
+		return 0, 0, errors.New("icmp code first is missing")
```

Comment:
Looks like this is applicable not only for the `icmp_code_*`, so probably it would be better to return a `${key1} is missing` error here.

Reply:
done